classes/boardmanager.php

```diff
@@ -0,0 +1,1041 @@
+                get_string('done', 'kanban') => '{"autoclose": true}',
+            ];
+            $columnids = [];
+            foreach ($columnnames as $columnname => $options) {
+                $columnids[] = $DB->insert_record('kanban_column', [
+                    'title' => $columnname,
+                    'sequence' => '',
+                    'kanban_board' => $boardid,
+                    'options' => $options,
+                    'timecreated' => time(),
+                    'timemodified' => time(),
+                ]);
+            }
+            $DB->update_record('kanban_board', ['id' => $boardid, 'sequence' => join(',', $columnids)]);
+            helper::update_cached_board($boardid);
+            return $boardid;
+        } else {
+            $fs = get_file_storage();
+            $template = get_cached_board($templateid);
+
+            // If it is a site wide template, we need system context to copy files.
+            if ($template->kanban_instance == 0) {
+                $context = context_system::instance(0);
+            } else {
+                $context = context_module::instance($this->cmid, 'kanban');
+            }
+
+            $newboard = (array)$template;
+            // By default, new board is not a template (can be overriden via $data).
+            $newboard['template'] = 0;
+            $newboard['timecreated'] = time();
+            $newboard['timemodified'] = time();
+            $newboard['userid'] = 0;
+            $newboard['group'] = 0;
+            unset($newboard['id']);
+
+            $newboard = array_merge($newboard, $data);
+
+            $newboard['id'] = $DB->insert_record('kanban_board', $newboard);
+            $columns = $DB->get_records('kanban_column', ['kanban_board' => $template->id]);
+            $cards = $DB->get_records('kanban_card', ['kanban_board' => $template->id]);
+            $newcolumn = [];
+            $newcard = [];
+            foreach ($columns as $column) {
+                $newcolumn[$column->id] = clone $column;
+                $newcolumn[$column->id]->kanban_board = $newboard['id'];
+                $newcolumn[$column->id]->timecreated = time();
+                $newcolumn[$column->id]->timemodified = time();
+                unset($newcolumn[$column->id]->id);
+                $newcolumn[$column->id]->id = $DB->insert_record('kanban_column', $newcolumn[$column->id]);
+            }
+            foreach ($cards as $card) {
+                $newcard[$card->id] = clone $card;
+                $newcard[$card->id]->kanban_board = $newboard['id'];
+                $newcard[$card->id]->timecreated = time();
+                $newcard[$card->id]->timemodified = time();
+                $newcard[$card->id]->kanban_column = $newcolumn[$card->kanban_column]->id;
+                unset($newcard[$card->id]->id);
+                // Remove user id of original creator.
+                unset($newcard[$card->id]->createdby);
```

Comment:
so it is ok that a card has no creator? Isn't this problematic?

Maybe the user importing the template should be the new creator?

Reply:
There is no problem when no creator is set. This field is only used to let users that have only addcard capability edit their own cards.